pkg/lib/cockpit-components-dropdown.tsx

```diff
@@ -35,49 +34,110 @@ import { EllipsisVIcon } from '@patternfly/react-icons';
+ *
+ * - "containsDrilldown" must be true
+ * - "id" must be given, and all DrilldownMenus also must have "id"s.
+ * - The MenuItems that navigate between the drilldows (i.e. the ones
+ *   with a "direction" attribute) must have a itemId that starts with
+ *   "drilldown:", and no other MenuItem must have a itemId that
+ *   starts with "drilldown:".
  */
-export const KebabDropdown = ({ dropdownItems, position = "end", isDisabled = false, toggleButtonId, isOpen, setIsOpen } : {
+export const KebabDropdown = ({ dropdownItems, position = "end", popperProps, isDisabled = false, toggleButtonId, id, isOpen, setIsOpen, containsDrilldown = false } : {
     dropdownItems: React.ReactNode,
     position?: DropdownPopperProps['position'],
+    popperProps?: DropdownPopperProps,
     isDisabled?: boolean,
-    toggleButtonId?: string;
+    toggleButtonId?: string,
+    id?: string,
     isOpen?: boolean, setIsOpen?: React.Dispatch<React.SetStateAction<boolean>>,
+    containsDrilldown?: boolean,
 }) => {
     const [isKebabOpenInternal, setKebabOpenInternal] = useState(false);
     const isKebabOpen = isOpen ?? isKebabOpenInternal;
     const setKebabOpen = setIsOpen ?? setKebabOpenInternal;
 
+    const [drilledInMenus, setDrilledInMenus] = React.useState<string[]>([]);
+    const [drilldownItemPath, setDrilldownItemPath] = React.useState<string[]>([]);
+    const [menuHeights, setMenuHeights] = React.useState<{[id: string]: number}>({});
+    const [activeMenu, setActiveMenu] = React.useState<string>(id || "");
+
+    function onDrillIn(_event: React.KeyboardEvent | React.MouseEvent, fromId: string, toId: string, pathId: string) {
+        setDrilledInMenus([...drilledInMenus, fromId]);
+        setDrilldownItemPath([...drilldownItemPath, pathId]);
+        setActiveMenu(toId);
+    }
+
+    function onDrillOut(_event: React.KeyboardEvent | React.MouseEvent, toId: string) {
+        const drilledInMenusSansLast = drilledInMenus.slice(0, drilledInMenus.length - 1);
+        const pathSansLast = drilldownItemPath.slice(0, drilldownItemPath.length - 1);
+        setDrilledInMenus(drilledInMenusSansLast);
+        setDrilldownItemPath(pathSansLast);
+        setActiveMenu(toId);
+    }
+
+    function resetDrilldown() {
+        setDrilledInMenus([]);
+        setDrilldownItemPath([]);
+        setActiveMenu(id || "");
+    }
+
+    function onGetMenuHeight(menuId: string, height: number) {
+        if (menuHeights[menuId] === undefined || (menuId !== id && menuHeights[menuId] !== height)) {
+            setMenuHeights({ ...menuHeights, [menuId]: height });
+        }
+    }
+
+    function onSelect(_event: React.MouseEvent<Element, MouseEvent> | undefined, itemId: string | number | undefined) {
+        if (!itemId || typeof itemId == "number" || !itemId.startsWith("drilldown:")) {
+            setKebabOpen(false);
+            resetDrilldown();
+        }
+    }
+
     return (
         <Dropdown
+            {... id ? { id } : { } }
```

Comment:
I thought the standard pattern was `...id && { id }`, I assume that stops the coverage complaint.

Reply:
Good point! So `{ ...false }` is ok for JSX? I'll change this.